transport.go

```diff
@@ -202,6 +224,56 @@ func (t *transport) Dial(ctx context.Context, raddr ma.Multiaddr, p peer.ID) (tp
 	return conn, nil
 }
 
+func (t *transport) holePunch(ctx context.Context, p peer.ID, network string, addr *net.UDPAddr) error {
+	pconn, err := t.connManager.Dial(network, addr)
+	if err != nil {
+		return err
+	}
+	defer pconn.DecreaseCount()
+
+	ctx, cancel := context.WithTimeout(ctx, time.Second)
+
+	hpkey := addr.String()
+	t.holePunchingMx.Lock()
+	t.holePunching[hpkey] = cancel
+	t.holePunchingMx.Unlock()
+
+	defer func() {
+		t.holePunchingMx.Lock()
+		defer t.holePunchingMx.Unlock()
+
+		_, ok := t.holePunching[hpkey]
+		if ok {
+			cancel()
+			delete(t.holePunching, hpkey)
+		}
```

Comment:
... then this block simplifies to just the `delete`.

Reply:
yes, but it will be called twice which panics!